server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -49,188 +51,314 @@
-            params.getReplicationManager(),
-            tier,
-            totalReplicantsInCluster,
-            expectedReplicantsInTier,
-            totalReplicantsInTier,
-            strategy,
-            serverHolderList,
-            segment
+      final ServerHolder candidate = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
+      if (candidate == null) {
+        log.warn(
+            "No available [%s] servers or node capacity to assign primary segment[%s]! " +
+            "Expected Replicants[%d]",
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
-        stats.addToTieredStat(ASSIGNED_COUNT, tier, assignedCount);
-        totalReplicantsInCluster += assignedCount;
+      } else {
+        // cache the result for later use.
+        strategyCache.put(tier, candidate);
+        if (topCandidate == null ||
+            candidate.getServer().getPriority() > topCandidate.getServer().getPriority()) {
+          topCandidate = candidate;
+        }
       }
+    }
 
-      loadStatus.put(tier, expectedReplicantsInTier - loadedReplicantsInTier);
+    if (topCandidate != null) {
+      // remove tier for primary replica
+      strategyCache.remove(topCandidate.getServer().getTier());
+      topCandidate.getPeon().loadSegment(segment, null);
     }
-    // Remove over-replication
-    stats.accumulate(drop(loadStatus, segment, params));
 
+    return topCandidate;
+  }
 
-    return stats;
+  /**
+   * @param tierToSkip if not null, this tier will be skipped from doing assignment, use when primary replica was
+   *                   assgined.
+   */
+  private void assignReplicas(
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment,
+      final CoordinatorStats stats,
+      @Nullable final String tierToSkip
+  )
+  {
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+      if (tier.equals(tierToSkip)) {
+        continue;
+      }
+      final int numAssigned = assignReplicasForTier(
+          tier,
+          entry.getIntValue(),
+          currentReplicants.getOrDefault(tier, 0),
+          params,
+          getFilteredHolders(tier, params.getDruidCluster(), createLoadQueueSizeLimitingPredicate(params)),
+          segment
+      );
+      stats.addToTieredStat(ASSIGNED_COUNT, tier, numAssigned);
+    }
   }
 
-  private int assign(
-      final ReplicationThrottler replicationManager,
+  private int assignReplicasForTier(
       final String tier,
-      final int totalReplicantsInCluster,
-      final int expectedReplicantsInTier,
-      final int totalReplicantsInTier,
-      final BalancerStrategy strategy,
-      final List<ServerHolder> serverHolderList,
+      final int targetReplicantsInTier,
+      final int currentReplicantsInTier,
+      final DruidCoordinatorRuntimeParams params,
+      final List<ServerHolder> holders,
```

Comment:
This is modified in the method call, suggest calling that out in the method docs

Reply:
I have moved the retrieval of `ServerHolder`s into the method to maintain "immutability" of parameters.